src/core/net/ip6_mpl.cpp

```diff
@@ -105,38 +102,265 @@ ThreadError Mpl::ProcessOption(const Message &aMessage, const Address &aAddress,
+{
+    int8_t diff;
+    MplBufferedMessageMetadata messageMetadata;
+
+    Message *message = mBufferedMessageSet.GetHead();
+    Message *nextMessage = NULL;
+
+    // Check if multicast forwarding is enabled.
+    VerifyOrExit(GetTimerExpirations() > 0,;);
+
+    while (message != NULL)
+    {
+        nextMessage = message->GetNext();
+        messageMetadata.ReadFrom(*message);
+
+        if (messageMetadata.GetSeedId() == aSeedId)
+        {
+            diff = static_cast<int8_t>(aSequence - messageMetadata.GetSequence());
+
+            if (diff > 0)
+            {
+                // Stop retransmitting MPL Data Message that is consider to be old. The new
+                // MPL Data Message will be added after function exits.
+                mBufferedMessageSet.Dequeue(*message);
+                message->Free();
+            }
+
+            break;
+        }
+
+        message = nextMessage;
+    }
+
+exit:
+    return;
+}
+
+void Mpl::AddBufferedMessage(Message &aMessage, uint16_t aSeedId, uint8_t aSequence, bool forward)
+{
+    uint32_t now = Timer::GetNow();
+    ThreadError error = kThreadError_None;
+    MplBufferedMessageMetadata messageMetadata;
+    Message *messageCopy = NULL;
+    uint8_t hopLimit;
+    uint32_t nextTransmissionTime;
+
+    VerifyOrExit((messageCopy = aMessage.Clone()) != NULL, error = kThreadError_NoBufs);
+
+    if (forward)
+    {
+        aMessage.Read(Header::GetHopLimitOffset(), Header::GetHopLimitSize(), &hopLimit);
+        VerifyOrExit(--hopLimit > 0, error = kThreadError_Drop);
+        messageCopy->Write(Header::GetHopLimitOffset(), Header::GetHopLimitSize(), &hopLimit);
+    }
+
+    messageMetadata.SetSeedId(aSeedId);
+    messageMetadata.SetSequence(aSequence);
+    messageMetadata.SetTransmissionCount(forward ? 0 : 1);
+
+    // Emulate Trickle timer behavior and set up the next retransmission within [0,I] range.
+    messageMetadata.SetTransmissionTime(now + (otPlatRandomGet() % (kDataMessageInterval + 1)));
+
+    // Append the message with MplBufferedMessageMetadata and add it to the queue.
+    SuccessOrExit(error = messageMetadata.AppendTo(*messageCopy));
+    mBufferedMessageSet.Enqueue(*messageCopy);
+
+    if (mRetransmissionTimer.IsRunning())
+    {
+        // If timer is already running, check if it should be restarted with earlier fire time.
+        nextTransmissionTime = mRetransmissionTimer.Gett0() + mRetransmissionTimer.Getdt();
+
+        if (messageMetadata.IsEarlier(nextTransmissionTime))
+        {
+            mRetransmissionTimer.Start(messageMetadata.GetTransmissionTime() - now);
+        }
+    }
+    else
+    {
+        // Otherwise just set the timer.
+        mRetransmissionTimer.Start(messageMetadata.GetTransmissionTime() - now);
+    }
+
+exit:
+
+    if (error != kThreadError_None && messageCopy != NULL)
+    {
+        messageCopy->Free();
+    }
+
+    return;
+}
+
+ThreadError Mpl::ForwardMessage(Message &aMessage)
```

Comment:
Stepping back a bit, do we need to special case locally-generated datagrams?  Can we just have `ProcessOption` handle both received and locally generated MPL data messages?

The reason I ask is that OpenThread also needs to handle externally generated IPv6 datagrams (e.g. from a linux or windows network stack) passed in view `otSendIp6Datagram()`.  It would be nice to only have one point of handling for all realm-local or larger scope multicast messages.